stripe/src/main/java/com/stripe/android/EphemeralKeyManager.java

```diff
@@ -0,0 +1,108 @@
+package com.stripe.android;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.VisibleForTesting;
+
+import java.lang.ref.WeakReference;
+import java.util.Calendar;
+import java.util.concurrent.TimeUnit;
+
+class EphemeralKeyManager {
+
+    private @Nullable EphemeralKey mEphemeralKey;
+    private @NonNull EphemeralKeyProvider mEphemeralKeyProvider;
+    private @Nullable Calendar mOverrideCalendar;
+    private @NonNull KeyManagerListener mListener;
+    private final long mTimeBufferInSeconds;
+
+    EphemeralKeyManager(
+            @NonNull EphemeralKeyProvider ephemeralKeyProvider,
+            @NonNull KeyManagerListener keyManagerListener,
+            long timeBufferInSeconds,
+            @Nullable Calendar overrideCalendar) {
+
+        mEphemeralKeyProvider = ephemeralKeyProvider;
+        mListener = keyManagerListener;
+        mTimeBufferInSeconds = timeBufferInSeconds;
+        mOverrideCalendar = overrideCalendar;
+        retrieveEphemeralKey();
+    }
+
+    void retrieveEphemeralKey() {
+        if (mEphemeralKey != null &&
+                !shouldRefreshKey(
+                        mEphemeralKey,
+                        mTimeBufferInSeconds,
+                        mOverrideCalendar)) {
+            mListener.onKeyUpdate(mEphemeralKey);
+            return;
+        }
+
+        mEphemeralKeyProvider.createEphemeralKey(
+                StripeApiHandler.API_VERSION,
+                new ClientKeyUpdateListener(this));
+    }
+
+    @Nullable
+    @VisibleForTesting
+    EphemeralKey getEphemeralKey() {
+        return mEphemeralKey;
+    }
+
+    private void updateSessionKey(@NonNull String key) {
```

Comment:
i wanted to emphasize whose key we're updating. It's actually not specifically talking to a CustomerSession, so maybe it should be a listener.